dmipy/utils/invariants.py

```diff
@@ -0,0 +1,576 @@
+        start = 0
+        for l in range(0, sh_order + 1, 2):
+            stop = start + 2*l+1
+            new_coef[...,start:stop] = new_coef[...,start:stop][...,::-1]
+            start += 2*l+1
+
+    return new_coef
+
+def _kappa(l2, l3, m2, m3):
+    r = max(abs(l2-l3), min(abs(m2+m3), abs(m2-m3)))
+    return r
+
+def _condition(l1, l2, l3, m1, m2, m3):
+    m_cond = False
+    l_cond = False
+    n_cond = False
+    if m1 in [m2+m3, m2-m3,-m2+m3, -m2-m3 ]:
+        m_cond=True
+    l_max = l2+l3
+
+    k= _kappa(l2,l3,m2,m3)
+    if (k+l_max) %2 ==0:
+        l_min = k
+    else:
+        l_min = k+1
+
+    bl = (l1+l2+l3) % 2 ==0
+
+    bl1 =False
+    for ll in range(l_min,l_max+1,2):
+        if l1==ll:
+            bl1 = True
+
+    l_cond = bl and bl1
+
+    bm = 0
+    bm += m1 <0
+    bm += m2 <0
+    bm += m3 <0
+
+    if bm %2 == 0:
+        n_cond = True
+
+    r =  m_cond and l_cond and n_cond
+
+    return r
+
+def _get_gaunt_index(l1,l2,l3,m1,m2,m3, sh_order = 16):
+    # n_c = (sh_order+1)*(sh_order+2)//2
+    # i1 = ((l1-1) * l1)//2
+    # i2 = ((l2-1) * l2)//2
+    # i3 = ((l3-1) * l3)//2
+    n_c = (sh_order+1)**2
+    i1 = l1**2
+    i2 = l2**2
+    i3 = l3**2
+    return (n_c**2) * (i1 + l1 + m1) + n_c*(i2 + l2 + m2) + i3 + l3 + m3
+
+def real_gaunt(l1,  l2,  l3,  mu1,  mu2,  mu3,  RG):
+    r"""Returns the real SH gaunt coefficients, namely the integral of
+    three SH with the same arguments
+    """
+    c = _get_gaunt_index(l1,l2,l3,mu1,mu2,mu3)
+    return RG[c]
+
+def real_gaunt4(l1,  l2,  l3,  l4,  mu1,  mu2,  mu3,  mu4,  RG):
+    r"""Returns the four SH real gaunt coefficients, namely the integral of
+    four SH with the same arguments
+    """
+    N = 2*np.max([l1,l2,l3,l4])
+    r  = 0
+
+    for l in range(0,N+1,1):
+        for m in range(-l,l+1):
+            c1 = _get_gaunt_index(l,l2,l3,m,mu2,mu3)
+            c2 = _get_gaunt_index(l1,l,l4,mu1,m,mu4)
+            #if condition(l,l2,l3,m,mu2,mu3) and condition(l1,l,l4,mu1,m,mu4):
+            r += RG[c1] * RG[c2]
+
+    return r
+
+def real_gaunt5(l1,  l2,  l3,  l4, l5,  mu1,  mu2,  mu3,  mu4, mu5, RG):
+    r"""Returns the five SH real gaunt coefficients, namely the integral of
+    five SH with the same arguments
+    """
+    N = 2*np.max([l1,l2,l3,l4,l5])
+    r  = 0
+    for l in range(0,N+1,1):
+        for m in range(-l,l+1):
+            if  _condition(l,l1,l2,m,mu1,mu2):
+                c1 = _get_gaunt_index(l,l1,l2,m,mu1,mu2)
+                for lp in range(0,N+1,1):
+                    for mp in range(-lp,lp+1):
+                        if  _condition(lp,l3,l4,mp,mu3,mu4) and _condition(l,lp,l5,m,mp,mu5):
+                            c2 = _get_gaunt_index(lp,l3,l4,mp,mu3,mu4)
+                            c3 = _get_gaunt_index(l,lp,l5,m,mp,mu5)
+                            r += RG[c1] * RG[c2] * RG[c3]
+
+    return r
+
+def get_invariants_list(sh_order):
+    r""" Returns the list of algebraically independent invariants for a given
```

Comment:
Perhaps some more information on what these number combinations are and where they come from.